Privacy preferences settings plugin

oguzeroglu opened the pull request:
Closes #1543

This PR does the same thing with #1586 however, @barmac and I decided that the App.js growing bigger as the new features are added is concerning so I re-implemented the same feature as a built-in plugin here.

nikku commented on client/src/app/primitives/__tests__/ModalSpec.js:
```diff
@@ -53,6 +53,25 @@ describe('<Modal>', function() {
 
+  describe('hideCloseIcon parameter', function() {
+
+    it('should render close icon if hideCloseIcon non existent', function() {
```
Note that we should also disable the close on escape and close on backdrop functionality if we hide the close icon.

Three ways exists how the user can escape normal modals:

* press escape
* click backdrop (grey overlay)
* press close button

Given that it is not only about hiding the close button, won't be "noUserEscape" or "confirm" (=> user must accept or reject using modal provided buttons) a better name for the prop?
We should probably extract that new infrastructure / Modal feature into its own commit.

oguzeroglu replied:
This feature also disables escape/backgroup click related modal escapes so I think you're right about the naming. "noUserEscape" seems like a good naming for this.

The noUserEscape feature is already in a separated commit: https://github.com/camunda/camunda-modeler/pull/1597/commits/c27d749c449233e8c82f42f232f9d2cf70f6a237

Did you mean something else?

oguzeroglu replied:
I actually now realized that ESC/Background click does not work because we don't pass an onClose method to the modal, so I think naming is correct here. We only want to hide the escape icon visually.

nikku replied:
Why can't we be consistent then => no `onClose` method => no `X` being shown?

nikku commented on client/src/plugins/privacy-preferences/PrivacyPreferencesView.js:
```diff
@@ -0,0 +1,122 @@
+    } = this.props;
+
+    return (
+      <Modal className={ css.View } hideCloseIcon={ true }>
```
This is the place where I'd assume => no `onClose` => no close button being shown.

`onClose` passed => standard modal interactions work => `ESC`, backdrop click

oguzeroglu replied:
This is a nice solution, however Modal.js has a default value for onClose: https://github.com/camunda/camunda-modeler/blob/c27d749c449233e8c82f42f232f9d2cf70f6a237/client/src/app/primitives/modal/Modal.js#L79

So we cannot get away with `if(!this.props.onClose)`, we can compare the function with noop function though, but I'm not exactly sure if that'd be semantically correct.

So we should either remove the default value or pass the hideCloseIcon parameter, I'd say lets keep the hideCloseIcon parameter. What do you say?

nikku replied:
Let's not think on the level as low as "our code is like this, we cannot change it" here. The fact that there was a noop in the past should not keep us from having a better API in the future.

What do we loose by making showing the close icon dependent on the onClose property? I guess not much. Showing a close button when the modal is not closable (and won't close) is a bug, not a feature.

On the other hand, in terms of API of our Modal: Should we as developers take two things into account (hideCloseButton, not adding onClick) to make it a confirmation dialog?

nikku commented on client/src/app/App.js:
```diff
@@ -118,8 +118,14 @@ export class App extends PureComponent {
 
     const userPlugins = this.getPlugins('client');
 
+    let defaultPluginsAry = this.props.hasServerInteraction ? defaultPlugins : (
```
What does _Ary_ stand fore here? This should also be a `const` I guess.

oguzeroglu replied:
This is now `const selectedDefaultPlugins`

nikku commented on client/src/app/AppParent.js:
```diff
@@ -397,6 +399,7 @@ export default class AppParent extends PureComponent {
         onReady={ this.handleReady }
         onError={ this.handleError }
         onWarning={ this.handleWarning }
+        hasServerInteraction={ hasServerInteraction }
```
We do not have to pass the flags here but could also directly use them:

```
import Flags from `...`

if (Flags.get('....') {

}
```

This way it is possible to check for the flag inside the plug-in and require no knowledge of the app.

nikku replied:
It's not like that this is a better way to do it, just wanted to make sure you've had that one on your radar when implementing the solution.

oguzeroglu replied:
Thanks!

nikku commented on client/src/plugins/privacy-preferences/PrivacyPreferences.js:
```diff
@@ -0,0 +1,78 @@
+
+const CONFIG_KEY = 'editor.privacyPreferences';
+
+export default class PrivacyPreferences extends PureComponent {
```
The idea here would be that this component considers the value of FLAGS(SERVER_INTERACTION) and becomes a noop.

oguzeroglu replied:
Thanks for the info. IMO it's better not to instantiate the class at all if the server-interaction flag is non-existent. Maybe if we really want to have less code in App.js, we can do some sort of early return inside PrivacyPreferences class depending on the flag, but I see no other benefit of doing that.

oguzeroglu replied:
I was thinking about doing the same for UpdateCheck module, which I also want to implement as a plugin.

oguzeroglu replied:
A benefit I can think of having certain plugins filtered out inside the App is that, in that case the render/constructor/mount methods of these plugins are not executed at all.

nikku commented on client/src/app/primitives/modal/Modal.js:
```diff
@@ -53,13 +53,14 @@ export default class Modal extends PureComponent {
       <div className={ css.ModalOverlay } onClick={ this.handleBackgroundClick }>
         <div className={ classNames(css.ModalContainer, className) } ref={ this.modalRef }>
-          <Close onClick={ this.close } />
+          { onClose && (<Close onClick={ this.close } />) }
```
This does break background clicks now, right? `this.close` does always assume the property is there.

oguzeroglu replied:
Ah yep, it did for the PrivacyPreferencesSettings, fixed it now.

barmac commented:
I pushed additional two commits:
* https://github.com/camunda/camunda-modeler/pull/1597/commits/9c19dcf8f9edebdc295b635b9a33cdb4d3b30ba5 restores the outline if the checkbox is focused; it's an accessibility feature and I can't see reason to disable this
* https://github.com/camunda/camunda-modeler/pull/1597/commits/c06bc917e198c232c4e1982dadfa929b22badc7d re-structures the form fields so that the fields are laid out evenly, we don't use excessive `div`s and we make use of the label accessibility functionality (cf. GIF below)

![Screen-Recording-2019-11-25-at-11 07 53](https://user-images.githubusercontent.com/28307541/69532187-95f9fc80-0f75-11ea-9b26-353a9674cfb0.gif)

oguzeroglu commented on client/src/plugins/privacy-preferences/PrivacyPreferences.js:
```diff
@@ -0,0 +1,86 @@
+
+  async componentDidMount() {
+
+    const hasServerInteraction = !!Flags.get(SERVER_INTERACTION) || process.env.NODE_ENV === 'test';
```
I added process.env.NODE_ENV here as overwriting Flags.get affects other tests. The other solution would be to overwrite it before the PrivacyPrefernecesSpec and revert it after the test however I'd rather just add this line here than to do that. Open to feedbacks.

oguzeroglu replied:
Ah, that's actually a bad idea as process does not exist in the client side. I thought Atom provided that class.

oguzeroglu commented:
@barmac I squashed your commits and removed the server-interaction flags related stuff from App.js and AppParent.js as Nico suggested. Also, see the comment above, let me know if you have any feedback on that.

barmac commented:
I pushed https://github.com/camunda/camunda-modeler/pull/1597/commits/5fd0a5d20f3e921d7d6d7d5e7e93b4d845241359 which hides the component completely behind a feature flag. The `componentDidMount` solution only prevented the config read operation but the component was still there. Now we simply return NoopComponent in the constructor.

oguzeroglu commented:
The componendDidMount solution would prevented the config read and subscribing to the menu action so it'd return an empty <Fragment> anyway. The test solution looks better, and it's good that we also got rid of that static variable which was useless now. I'll squash it.

oguzeroglu commented:
I don't like the comment message though: Hide **correctly** behind the feature flag. What's right and what's wrong here? The **right** solution is to never initialize the Component in the first place, which was the first solution. But as we do not want to add more stuff to App class, we discarded that. IMO returning another class from a constructor is not the **right** solution, it seems even a bit hacky to me.

But yeah lets squash that one too anyway :)

barmac commented:
Dunno, it's just a proposal aimed at not rendering `PrivacyPreferences` at all. 

To be honest I first tried out the `export default Flag.get(...) ? PrivacyPreferences : NoopComponent` approach, but it does not work due to the Flags beeing initialized after the imports finish. We could experiment with the dynamic imports, but then it appears that the whole `constructor`-based logic has to be rewritten due to required `Promise`  support.

oguzeroglu commented:
@barmac Squashed your commit!

barmac commented on client/src/plugins/privacy-preferences/constants.js:
```diff
@@ -0,0 +1,47 @@
+
+export const PRIVACY_TEXT_FIELD = 'To enhance user experience, Camunda Modeler can integrate with 3rd party services, which requires external network requests. Please choose from the settings below.';
+
+export const PRIVACY_POLICY_URL = 'https://camunda.com/legal/privacy/';
```
This could actually be read from the env and substituted during the build.

oguzeroglu replied:
I think it's ok to keep it here. Otherwise, we would need to write a new provider to config to provide this env variable to frontend right? Isn't it much easier like this?

nikku replied:
Let's keep it hard coded for now. We need to make this plugable only once customizing of the update server endpoint becomes a requirement.
